play/psycopg3/async_tail.py

```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""async_tail
+
+Module showing an example for how to asynchronously TAIL a Materialized view.
+"""
+
+import argparse
+import asyncio
+
+import psycopg3
+
+
+async def tail_view(args):
+    """Continuously print changes to a Materialize View."""
+    dsn = f"postgresql://{args.host}:{args.port}/materialize"
+    async with await psycopg3.AsyncConnection.connect(dsn) as conn:
+        async with await conn.cursor() as cursor:
+            query = f"COPY (TAIL {args.view} WITH (PROGRESS)) TO STDOUT"
+            async with cursor.copy(query) as tail:
+                async for row in tail:
+                    row = row.decode("utf-8")
+                    (timestamp, progressed, diff, *columns) = row.strip().split("\t")
+                    print(f"{timestamp} {progressed} {diff} {columns}")
```

Comment:
we should add an issue to return to this when the tail timeout code matt is currently working on lands and we can stop parsing the output, instead getting it as real rows.

Reply:
That's always going to be less efficient I think than using the copy protocol. I think the current plan is to see if we can get the psycopg3 developers to build support for binary copy.

Reply:
That seems like a fine better issue to create. The main point from my PoV is that `\t` is a bad thing to need to split on.

Reply:
Turns out I was misunderstanding the copy protocol, I think this would be more clear if it demonstrated that the columns will contain escaped characters. Perhaps by sticking this in between the split and the print:

```python
columns = [c.decode('unicode_escape') for c in columns]
```

Reply:
That's probably subtly incorrect, based on this SO post: https://stackoverflow.com/a/24519338


I'd rather either leave it as is with a comment like `TODO: handle unescaping`, or do the exactly correct thing, rather than doing something that is subtly wrong. This is just in "play", so we don't need to hold it to a particularly high standard until/unless we move it to "demo".

Reply:
Here's a plausibly correct unescape routine: https://repl.it/join/mrloaelz-benesch

Reply:
yeah that won't handle the `\xDD` or `\DDD` formats, either, so I agree we should just stick a TODO in there to point out that it is possible but not implemented.

Reply:
`COPY TO` will never output `\xDD` or `\DDD` escape codes.

Reply:
Ah, yeah, "presently". Despite how great the postgres docs are, I seem to always think they end before they do and missed that. Your function seems reasonable, I'm not sure if it needs to handle `\.`, which it seems like our COPY TO will probably also never emit?

Reply:
> which it seems like our COPY TO will probably also never emit

Yup, exactly.